server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -102,316 +36,50 @@ public static void removeThreadContext(ThreadContext threadContext) {
-        }
-        for (int i = 0x26; i <= 0x5B; i++) {
-            doesNotNeedEncoding.set(i);
-        }
-        for (int i = 0x5D; i <= 0x7E; i++) {
-            doesNotNeedEncoding.set(i);
-        }
-        for (int i = 0x80; i <= 0xFF; i++) {
-            doesNotNeedEncoding.set(i);
-        }
-        assert doesNotNeedEncoding.get('%') == false : doesNotNeedEncoding;
+    //TODO fix logger usage check
+    public DeprecationLoggerBuilder deprecate(final String key, final String msg, final Object... params) {
+        return new DeprecationLoggerBuilder()
+            .withDeprecation(key, msg, params);
     }
 
-    private static final Charset UTF_8 = StandardCharsets.UTF_8;
-
-    /**
-     * Encode a string containing characters outside of the legal characters for an RFC 7230 quoted-string.
-     *
-     * @param s the string to encode
-     * @return the encoded string
-     */
-    static String encode(final String s) {
-        // first check if the string needs any encoding; this is the fast path and we want to avoid creating a string builder and copying
-        boolean encodingNeeded = false;
-        for (int i = 0; i < s.length(); i++) {
-            int current = s.charAt(i);
-            if (doesNotNeedEncoding.get(current) == false) {
-                encodingNeeded = true;
-                break;
-            }
-        }
-
-        if (encodingNeeded == false) {
-            return s;
-        }
+    public class DeprecationLoggerBuilder {
 
-        final StringBuilder sb = new StringBuilder(s.length());
-        for (int i = 0; i < s.length();) {
-            int current = s.charAt(i);
-            /*
-             * Either the character does not need encoding or it does; when the character does not need encoding we append the character to
-             * a buffer and move to the next character and when the character does need encoding, we peel off as many characters as possible
-             * which we encode using UTF-8 until we encounter another character that does not need encoding.
-             */
-            if (doesNotNeedEncoding.get(current)) {
-                // append directly and move to the next character
-                sb.append((char) current);
-                i++;
-            } else {
-                int startIndex = i;
-                do {
-                    i++;
-                } while (i < s.length() && doesNotNeedEncoding.get(s.charAt(i)) == false);
+        private String deprecationKey;
+        private ESLogMessage deprecationMessage;
 
-                final byte[] bytes = s.substring(startIndex, i).getBytes(UTF_8);
-                // noinspection ForLoopReplaceableByForEach
-                for (int j = 0; j < bytes.length; j++) {
-                    sb.append('%').append(hex(bytes[j] >> 4)).append(hex(bytes[j]));
-                }
-            }
+        public DeprecationLoggerBuilder withDeprecation(String key, String msg, Object[] params) {
```

Comment:
This is a good idea. Still sometimes provided key might be useful - if someone calls deprecation logger twice but with different message. (can't find that in codebase quickly)

But do you think we should iterate all usages of DeprecationLogger and remove the use of `deprecate(key,message,params)` in favour of deprecate(message,params) ?

Reply:
another interesting idea might to instead of using a message as a key would be to use a `call site` as a key. 
a call site would be a place in code that is more unique then a message (unlikely but sitll)
this can be taken simply from a stacktrace from a newly created exception.
flogger does this https://github.com/google/flogger/blob/b081bc6f209065aea11a797db3ea833bed0ad4b3/api/src/main/java/com/google/common/flogger/util/CallerFinder.java#L43

but that should be a different refactoring

Reply:
I like the idea of eliminating the need of a key by caching by the call site. https://docs.oracle.com/javase/9/docs/api/java/lang/StackWalker.html might help here.  

also, agree future refactoring.